File: scripts/tui-smoke.py
#!/usr/bin/env python3
import argparse
import errno
import fcntl
import json
import os
import re
import select
import signal
import struct
import termios
import time
import tempfile
from pathlib import Path

parser = argparse.ArgumentParser()
default_root = str(Path(__file__).resolve().parents[1])
parser.add_argument("--root", default=default_root)
parser.add_argument("--cwd")
parser.add_argument("--home")
args = parser.parse_args()
if not args.cwd:
    args.cwd = tempfile.mkdtemp(prefix="cascade-tui-project-")
if not args.home:
    args.home = tempfile.mkdtemp(prefix="cascade-tui-home-")

env = os.environ.copy()
env.update({"HOME": args.home, "TERM": "xterm-256color", "PI_OFFLINE": "1"})
for key in list(env):
    if key.endswith("_API_KEY") or key in {"MODEL_API_KEY", "OPENROUTER_API_KEY"}:
        env.pop(key, None)

pid, fd = os.forkpty()
if pid == 0:
    os.chdir(args.cwd)
    os.execvpe("node", ["node", os.path.join(args.root, "bin", "cascade.mjs"), "--offline", "--no-session"], env)

fcntl.ioctl(fd, termios.TIOCSWINSZ, struct.pack("HHHH", 42, 150, 0, 0))
os.set_blocking(fd, False)
chunks = []

def drain(seconds):
    deadline = time.monotonic() + seconds
    while time.monotonic() < deadline:
        ready, _, _ = select.select([fd], [], [], 0.1)
        if not ready:
            continue
        try:
            data = os.read(fd, 65536)
            if not data:
                break
            chunks.append(data)
        except OSError as error:
            if error.errno not in {errno.EIO, errno.EAGAIN}:
                raise
            if error.errno == errno.EIO:
                break

def send(data):
    os.write(fd, data)

drain(2.5)
send(b"\x0f")  # ctrl+o, expand startup/help if available
drain(0.5)
send(b"/cascade\r")
drain(0.8)
send(b"/cascade-mode dual\r")
drain(1.0)
send(b"/cascade\r")
drain(0.8)
send(b"\x04")
drain(0.2)
send(b"\x04")
drain(1.0)

try:
    waited, status = os.waitpid(pid, os.WNOHANG)
    if waited == 0:
        os.kill(pid, signal.SIGTERM)
        time.sleep(0.2)
        waited, status = os.waitpid(pid, os.WNOHANG)
        if waited == 0:
            os.kill(pid, signal.SIGKILL)
            _, status = os.waitpid(pid, 0)
except ProcessLookupError:
    status = 0

raw = b"".join(chunks).decode("utf-8", "replace")
plain = re.sub(r"\x1b\][^\x07]*(?:\x07|\x1b\\)", "", raw)
plain = re.sub(r"\x1b\[[0-?]*[ -/]*[@-~]", "", plain).replace("\r", "")
report = {
    "hasCascadeBrand": "Cascade" in plain,
    "hasOldPiHeader": "pi v0.84.2" in plain.lower(),
    "hasPiContext": "PI_ONLY_CONTEXT_SENTINEL" in plain,
    "hasNativeModelHint": "/model · /cascade-setup" in plain,
    "hasClearSingleStatus": "Single · Worker:" in plain,
    "hasClearDualStatus": "Dual · Active Worker:" in plain and "Expert: on-demand" in plain,
    "hasCrypticStatus": bool(re.search(r"worker · (?:single|dual) · route ", plain)),
    "tail": plain[-7000:]
}
if (
    not report["hasCascadeBrand"]
    or report["hasOldPiHeader"]
    or report["hasPiContext"]
    or not report["hasNativeModelHint"]
    or not report["hasClearSingleStatus"]
    or not report["hasClearDualStatus"]
    or report["hasCrypticStatus"]
):
    print(json.dumps(report))
    raise SystemExit(1)
print(json.dumps(report))

File: scripts/tui-smoke.mjs
#!/usr/bin/env node
import { spawn, spawnSync } from "node:child_process";
import { existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

if (process.platform !== "linux") {
  console.log(`Cascade TUI smoke skipped on ${process.platform}; Linux PTY coverage runs in CI.`);
  process.exit(0);
}
if (spawnSync("script", ["--version"], { stdio: "ignore" }).error) {
  console.log("Cascade TUI smoke skipped because util-linux script(1) is unavailable.");
  process.exit(0);
}

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const cli = join(root, "bin", "cascade.mjs");
const workspace = mkdtempSync(join(tmpdir(), "cascade-real-tui-"));
const home = join(workspace, "home");
const project = join(workspace, "project");
mkdirSync(join(project, ".cascade"), { recursive: true });
mkdirSync(home, { recursive: true });
writeFileSync(join(project, "README.md"), "# TUI smoke\n", "utf8");
writeFileSync(join(project, ".cascade", "config.json"), `${JSON.stringify({
  schemaVersion: 1,
  mode: "single",
  worker: {
    useNativeModel: false,
    provider: "meta-model-api",
    model: "muse-spark-1.2",
    thinking: "medium"
  },
  expert: {
    useNativeModel: false,
    provider: "meta-model-api",
    model: "muse-spark-1.2",
    thinking: "high",
    tools: ["read", "grep", "find", "ls", "bash"]
  },
  routing: { autoConsult: false },
  privacy: { classification: "internal", allowContributor: false }
}, null, 2)}\n`, "utf8");

const quote = (value) => `'${String(value).replaceAll("'", "'\\''")}'`;
const command = `${quote(process.execPath)} ${quote(cli)} --no-session`;
const child = spawn("script", ["-qefc", command, "/dev/null"], {
  cwd: project,
  env: {
    ...process.env,
    HOME: home,
    XDG_CONFIG_HOME: join(home, ".config"),
    XDG_STATE_HOME: join(home, ".local", "state"),
    MODEL_API_KEY: "cascade-tui-smoke-key",
    CASCADE_PROJECT_TRUSTED: "1",
    TERM: "xterm-256color",
    CI: ""
  },
  stdio: ["pipe", "pipe", "pipe"]
});

let output = "";
child.stdout.setEncoding("utf8");
child.stderr.setEncoding("utf8");
child.stdout.on("data", (chunk) => { output += chunk; });
child.stderr.on("data", (chunk) => { output += chunk; });
const delay = (ms) => new Promise((resolvePromise) => setTimeout(resolvePromise, ms));
const send = async (value, wait = 350) => {
  child.stdin.write(value);
  await delay(wait);
};

let timeout;
try {
  timeout = setTimeout(() => child.kill("SIGKILL"), 30000);
  await delay(1800);

  // Prove native Pi menus remain reachable. Cascade must not shadow these.
  await send("/settings\r", 500);
  await send("\u001b", 300);
  await send("/model\r", 500);
  await send("\u001b", 300);
  await send("/login\r", 500);
  await send("\u001b", 300);

  // Complete the actual Cascade TUI setup: project scope, dual mode,
  // current worker/expert models, auto consultation, internal privacy, save.
  await send("/cascade-setup\r", 500);
  await send("\r", 300);                 // project scope
  await send("\u001b[B\r", 300);        // dual mode
  await send("\r", 300);                 // current Pi worker model
  await send("\r", 300);                 // keep worker thinking
  await send("\r", 300);                 // current Pi expert model
  await send("\r", 300);                 // keep expert thinking
  await send("\u001b[B\r", 300);        // enable auto consultation
  await send("\u001b[B\u001b[B\r", 300); // internal privacy
  await send("\r", 700);                 // confirm save

  await send("/cascade-role expert\r", 600);
  await send("/cascade-role worker\r", 600);
  await send("/settings\r", 500);
  await send("\u001b", 300);
  await send("\u0003", 500);
  await send("\u0003", 200);

  const exit = await Promise.race([
    new Promise((resolvePromise) => child.once("exit", (code, signal) => resolvePromise({ code, signal }))),
    delay(3000).then(() => ({ code: null, signal: "timeout" }))
  ]);
  if (exit.signal === "timeout") child.kill("SIGKILL");

  const configPath = join(project, ".cascade", "config.json");
  if (!existsSync(configPath)) throw new Error("TUI setup did not save .cascade/config.json");
  const saved = JSON.parse(readFileSync(configPath, "utf8"));
  if (saved.mode !== "dual") throw new Error(`TUI setup did not enable dual mode: ${saved.mode}`);
  if (!saved.worker?.provider || !saved.worker?.model) throw new Error("TUI setup did not persist the worker model");
  if (!saved.expert?.provider || !saved.expert?.model) throw new Error("TUI setup did not persist the expert model");
  if (saved.routing?.autoConsult !== true) throw new Error("TUI setup did not enable automatic expert consultation");
  if (saved.privacy?.classification !== "internal") throw new Error("TUI setup did not persist repository privacy");

  const plain = output.replace(/\u001b\[[0-?]*[ -/]*[@-~]/g, "").replace(/\r/g, "");
  if (/TypeError|ReferenceError|SyntaxError|Unhandled|ERR_/.test(plain)) {
    throw new Error(`TUI emitted a runtime error:\n${plain.slice(-5000)}`);
  }
  for (const commandName of ["/settings", "/model", "/login", "/cascade-setup"]) {
    if (!plain.includes(commandName)) throw new Error(`TUI transcript did not contain ${commandName}`);
  }
  console.log("Real Cascade TUI smoke passed: native Pi menus, dual setup save, and role switching.");
} finally {
  clearTimeout(timeout);
  if (!child.killed) child.kill("SIGKILL");
  if (process.env.CASCADE_TUI_SMOKE_KEEP !== "1") rmSync(workspace, { recursive: true, force: true });
  else console.log(`TUI smoke workspace retained at ${workspace}`);
}
